src/main/java/net/fabricmc/loader/entrypoint/EntrypointContainerImpl.java

```diff
@@ -16,21 +16,28 @@
 
 package net.fabricmc.loader.entrypoint;
 
-import net.fabricmc.loader.ModContainer;
+import net.fabricmc.loader.api.ModContainer;
 import net.fabricmc.loader.api.entrypoint.EntrypointContainer;
 
+import java.util.function.Supplier;
+
 public class EntrypointContainerImpl<T> implements EntrypointContainer<T> {
 	private final ModContainer container;
-	private final T entrypoint;
+	private final Supplier<T> entrypointSupplier;
+	private T instance;
 
-	public EntrypointContainerImpl(ModContainer container, T entrypoint) {
+	public EntrypointContainerImpl(ModContainer container, Supplier<T> entrypointSupplier) {
 		this.container = container;
-		this.entrypoint = entrypoint;
+		this.entrypointSupplier = entrypointSupplier;
 	}
 
 	@Override
 	public T getEntrypoint() {
-		return entrypoint;
+		if (instance == null) {
+			this.instance = entrypointSupplier.get();
```

Comment:
This should be made thread safe, the easiest solution is making the entire method synchronized.